src/utils/async.js

```diff
@@ -0,0 +1,245 @@
+/* eslint-disable max-classes-per-file */
+import { MapWithCapacity } from './generic';
+
+/**
+ * A promise that can be reused.
+ */
+export class ReusablePromise {
+  /**
+   *
+   * @param {object} configuration
+   * @param {boolean} [configuration.save=false] - if set to `true`, stores the result of the first successful promise forever.
+   */
+  constructor({ save = false } = {}) {
+    this.promise = null;
+    this.save = save;
+  }
+
+  /**
+   * Checks if a promise exists, if so, `await`s it, otherwise inserts a new `Promise` produced by
+   * calling `createPromise` function.
+   *
+   * @template T
+   * @param {function(): Promise<T>} createPromise
+   * @returns {Promise<T>}
+   */
+  async call(createPromise) {
+    if (this.promise != null) {
+      return await this.promise;
+    }
+    const promise = createPromise();
+    this.promise = promise;
+
+    let res;
+    let err;
+    try {
+      res = await this.promise;
+    } catch (e) {
+      err = e;
+    } finally {
+      if ((!this.save || err != null) && this.promise === promise) {
+        this.clear();
+      }
+    }
+
+    if (err != null) {
+      throw err;
+    }
+    return res;
+  }
+
+  clear() {
+    this.promise = null;
+  }
+}
+
+/**
+ * A map where each entry represents a keyed promise that can be reused.
+ */
+export class ReusablePromiseMap {
+  /**
+   *
+   * @param {object} configuration
+   * @param {boolean} [configuration.save=false] - if set to `true`, stores the result of the first successful promise for each key.
+   * @param {?number} [configuration.capacity] - max capacity of the underlying container.
+   */
+  constructor({ capacity, save = false } = {}) {
+    this.map = capacity != null ? new MapWithCapacity(capacity) : new Map();
+    this.queue = [];
+    this.save = save;
+  }
+
+  /**
+   * Checks if a promise with the supplied key exists, if so, `await`s it, otherwise inserts a new `Promise` produced by
+   * calling `createPromise` function.
+   * In case capacity is reached, the current call will be paused until the next promise(s) is resolved.
+   *
+   * @template T
+   * @param {*} key
+   * @param {function(): Promise<T>} createPromise
+   * @returns {Promise<T>}
+   */
+  async callByKey(key, createPromise) {
+    if (this.map.has(key)) {
+      return await this.map.get(key);
```

Comment:
To build a correct stack, otherwise, in case of error, the `callByKey` method will be excluded from the stack trace. There was a rule forcing us to replace `return await` with `return` but it was deprecated, see https://eslint.org/docs/latest/rules/no-return-await.